test/topology_custom/test_aggregation.py

```diff
@@ -0,0 +1,61 @@
+#
+# Copyright (C) 2025-present ScyllaDB
+#
+# SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+#
+import asyncio
+import pytest
+import time
+import logging
+import random
+
+from test.pylib.manager_client import ManagerClient
+from test.pylib.rest_client import inject_error
+from test.pylib.util import wait_for, wait_for_cql_and_get_hosts
+
+logger = logging.getLogger(__name__)
+
+@pytest.mark.asyncio
+async def test_cancel_mapreduce(manager: ManagerClient):
+    """
+    This test verifies that stopping the supercoordinator of a mapreduce task cancels
+    outgoing queries to other nodes, which would otherwise prevent the shutdown.
+    """
+
+    [s1, s2] = await manager.servers_add(2, cmdline=["--logger-log-level", "forward_service=trace"])
+    cql = manager.get_cql()
+    hosts = await wait_for_cql_and_get_hosts(cql, [s1, s2], time.time() + 30)
+
+    [host1] = filter(lambda host: host.address == s1.ip_addr, hosts)
+    host_id2 = await manager.get_host_id(s2.server_id)
+
+    await cql.run_async("CREATE KEYSPACE ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1}")
+    await cql.run_async("CREATE TABLE ks.t (pk int PRIMARY KEY, v int)")
+
+    # Distribute data across the nodes.
+    for _ in range(100):
+        # Note: CQL int is a 32-bit integer.
+        pk = random.randint(-2**30, 2**30)
+        v = random.randint(-2**30, 2**30)
+        await cql.run_async(f"INSERT INTO ks.t (pk, v) VALUES ({pk}, {v})")
+
+    s1_log = await manager.server_open_log(s1.server_id)
+    s2_log = await manager.server_open_log(s2.server_id)
+
+    s1_mark = await s1_log.mark()
+    s2_mark = await s2_log.mark()
+
+    # Prevent finishing local mapreduce tasks on node 2.
+    async with inject_error(manager.api, s2.ip_addr, "prevent_local_mapreduce_tasks"):
+        # Make node 1 the supercoordinator of the mapreduce task corresponding to aggregation.
+        # We use this timeout because it's longer than the cumulative timeout of the following
+        # steps. For the test to be reliable, the query cannot end on its own and that's also
+        # why we're intentionally discarding the future here.
+        _ = cql.run_async("SELECT count(*) FROM ks.t BYPASS CACHE USING TIMEOUT 600s", host=host1)
```

Comment:
AFAIK runaway tasks are bound to cause problems in our test framework, we have some code that detects whether there are some tasks running at the end of the test and mark it as failed if there are some. Moreover, I'm not sure whether the query will run at all if you don't await it - IIRC Python coroutines are lazy and do nothing unless you poll them. You should make sure that you poll it and wait until it completes (and ignore the error).

I recommend taking a look into asyncio's [TaskGroups](https://docs.python.org/3/library/asyncio-task.html#task-groups), they would be a good fit here.